lib/algo/DMD.cpp

```diff
@@ -343,11 +395,98 @@ DMD::constructDMD(const Matrix* f_snapshots,
         d_S_inv->item(i, i) = 1 / d_factorizer->S[static_cast<unsigned>(i)];
     }
 
+    Matrix* Q = NULL;
+
+    // If W0 is not null, we need to ensure it is in the basis of W.
+    if (W0 != NULL)
+    {
+        CAROM_VERIFY(W0->numRows() == d_basis->numRows());
+        CAROM_VERIFY(linearity_tol >= 0.0);
+        std::vector<int> lin_independent_cols_W;
+
+        // Find which columns of d_basis are linearly independent from W0
+        for (int j = 0; j < d_basis->numColumns(); j++)
+        {
+            // l = W0' * u
+            Vector W_col(f_snapshots->numRows(), f_snapshots->distributed());
+            for (int i = 0; i < f_snapshots->numRows(); i++)
+            {
+                W_col.item(i) = d_basis->item(i, j);
+            }
+            Vector* l = W0->transposeMult(W_col);
```

Comment:
This allocates `l` for each `j`. It would be more efficient to construct a work vector before the loop. Also for `W0l`.

Reply:
Added.